core/src/main/java/io/stargate/core/activator/BaseActivator.java

```diff
@@ -0,0 +1,169 @@
+   * user-provided {@link this#createService(List)} and register it in the OSGi using {@link
+   * BundleContext#registerService(Class, Object, java.util.Dictionary)}.
+   */
+  @Override
+  public synchronized void start(BundleContext context) throws InvalidSyntaxException {
+    logger.info("Starting {} ...", activatorName);
+    this.context = context;
+
+    for (Class<?> dependentService : dependentServices) {
+      ServiceReference<?> serviceReference =
+          context.getServiceReference(dependentService.getName());
+      if (serviceReference == null) {
+        logger.debug(
+            "{} service is null, registering a listener to get notification when it will be ready.",
+            dependentService.getName());
+        context.addServiceListener(
+            this, String.format("(objectClass=%s)", dependentService.getName()));
+      } else {
+        // if the service is present, get it
+        registeredServices.put(dependentService, context.getService(serviceReference));
+      }
+    }
+
+    if (registeredServices.values().stream().anyMatch(Objects::isNull)) {
+      // It will be started once all dependent services are registered
+      return;
+    }
+
+    // all dependant services are present
+    startServiceInternal(new LinkedList<>(registeredServices.values()));
+  }
+
+  @Override
+  public synchronized void stop(BundleContext context) throws Exception {
+    // no-op
+  }
+
+  /**
+   * It will try to match all dependentServices with a ServiceEvent notification. If the
+   * notification for dependant service contains a non-null value, it will start the target service.
+   * After the notification is handled, it checks if all dependentServices are not null. If they
+   * are, the client's provided {@link this#createService(List)} is called, and the service is
+   * registered. It will not register the service if it was already registered in the {@link
+   * this#start(BundleContext)}.
+   */
+  @Override
+  public synchronized void serviceChanged(ServiceEvent event) {
+    for (Class<?> dependentService : dependentServices) {
+      Object registeredService = BundleUtils.getRegisteredService(context, event, dependentService);
+      if (registeredService != null) {
+        // capture registration only if the instance is not null
+        registeredServices.put(dependentService, registeredService);
+      }
+    }
+    if (registeredServices.values().stream().allMatch(Objects::nonNull)) {
+      startServiceInternal(new LinkedList<>(registeredServices.values()));
+    }
+  }
+
+  private synchronized void startServiceInternal(List<Object> dependentServices) {
+    if (started) {
+      logger.info("The {} is already started. Ignoring the start request.", activatorName);
+      return;
+    }
+    started = true;
+    ServiceAndProperties service = createService(dependentServices);
+    context.registerService(targetServiceClass.getName(), service.service, service.properties);
```

Comment:
Is there any particular reason to use the class _name_ for service registration as opposed to the class _object_?

Reply:
Also, what about things that don't need to be registered? For example, a top level service like graphql doesn't need to register itself.

Reply:
done in: https://github.com/stargate/stargate/pull/370/commits/47f882cccf8a281b5680d272bf5d8f372b86f0ec

Reply:
>
Is there any particular reason to use the class name for service registration as opposed to the class object?

the `registerService` API is not accepting the `targetServiceClass` of a `Class<?> type` as an argument.